File: frontend/lib/versions.ts
/**
 * Content versioning — types, Persian vocabulary and API wrappers.
 *
 * Field names are lifted verbatim from `backend/app/schemas/versions.py`
 * (ContentVersionListItem / ContentVersionRead / ContentVersionDiff /
 * ContentVersionSummary). Renaming anything here silently produces `undefined`
 * in the UI instead of a type error, because the responses arrive as plain JSON.
 *
 * There is no create/update type in here on purpose: revisions are minted by
 * the backend whenever an article changes, never posted by a client. The only
 * writable payload in the whole module is the rollback note.
 *
 * Note the two read shapes. The history list returns metadata only — no
 * `content_markdown` / `content_html` — so a 50-revision sidebar does not ship
 * a megabyte of duplicated article bodies. Bodies arrive only from
 * `getVersion()`.
 */

import { api } from "@/lib/api-client";

// ---------------------------------------------------------------- vocabularies

export type ContentChangeType =
  | "created"
  | "edited"
  | "ai_rewrite"
  | "rollback"
  | "published"
  | "imported";

export const CHANGE_TYPE_LABELS_FA: Record<string, string> = {
  created: "ایجاد شد",
  edited: "ویرایش دستی",
  ai_rewrite: "بازنویسی با هوش مصنوعی",
  rollback: "بازگردانی نسخه",
  published: "انتشار",
  imported: "درون‌ریزی",
};

/**
 * Mirrors CONTENT_SYSTEM_CHANGE_TYPES in the model. `changed_by` is null on
 * these, so the author cell shows "سیستم" rather than an empty space.
 */
export const SYSTEM_CHANGE_TYPES: string[] = ["ai_rewrite", "imported"];

export function isSystemChange(changeType: string | null | undefined): boolean {
  if (!changeType) return false;
  return SYSTEM_CHANGE_TYPES.includes(changeType);
}

/** Diff line kinds as normalised by the service's `_classify_diff_line`. */
export type DiffLineKind = "context" | "added" | "removed" | "header" | "hunk";

// -------------------------------------------------------------------- entities

export interface ContentVersionDiffStats {
  added_chars: number;
  removed_chars: number;
  added_words: number;
  removed_words: number;
}

/** History sidebar row — deliberately no bodies. */
export interface ContentVersionListItem {
  id: string;
  organization_id: string;
  website_id: string;
  article_id: string;
  version_number: number;
  title: string;
  seo_score: number;
  change_type: ContentChangeType | string;
  change_summary: string | null;
  changed_by: string | null;
  diff_stats: Record<string, number>;
  is_current: boolean;
  created_at: string;
}

/** Full snapshot, as returned by `GET /versions/{id}`. */
export interface ContentVersion extends ContentVersionListItem {
  content_markdown: string;
  content_html: string;
  seo_metadata: Record<string, any>;
}

export interface ContentVersionDiffLine {
  kind: DiffLineKind | string;
  text: string;
}

export interface ContentVersionDiff {
  article_id: string;
  from_version_number: number;
  to_version_number: number;
  from_created_at: string;
  to_created_at: string;
  title_changed: boolean;
  from_title: string;
  to_title: string;
  from_seo_score: number;
  to_seo_score: number;
  seo_score_delta: number;
  stats: ContentVersionDiffStats;
  lines: ContentVersionDiffLine[];
  identical: boolean;
  truncated: boolean;
}

export interface ContentVersionSummary {
  article_id: string;
  total_versions: number;
  current_version_number: number | null;
  last_changed_at: string | null;
  contributors: number;
}

// ------------------------------------------------------------------- requests

export interface ContentVersionRollbackRequest {
  change_summary?: string;
}

export interface ContentVersionRollbackResult {
  restored_from_version_number: number;
  new_version: ContentVersion;
}

// ----------------------------------------------------------------- formatting

/**
 * Renders diff_stats as a compact Persian summary, e.g. "۳۴۰ واژه افزوده، ۱۲
 * واژه حذف شده". Words, not characters: a Persian reader judges the size of an
 * edit in words, and character counts inflate wildly on RTL text with diacritics.
 *
 * Returns "بدون تغییر" when both sides are zero so a caller never has to decide
 * what an empty string should look like. Accepts the loose
 * `Record<string, number>` that the list rows carry, because JSONB round-trips
 * as an untyped object and an older row may predate a key.
 */
export function formatDiffStatsFa(
  stats: Record<string, number> | ContentVersionDiffStats | null | undefined
): string {
  if (!stats) return "بدون تغییر";
  const added = Number((stats as any).added_words ?? 0);
  const removed = Number((stats as any).removed_words ?? 0);
  if (!added && !removed) return "بدون تغییر";

  const parts: string[] = [];
  if (added) parts.push(`${added.toLocaleString("fa-IR")} واژه افزوده`);
  if (removed) parts.push(`${removed.toLocaleString("fa-IR")} واژه حذف شده`);
  return parts.join("، ");
}

/** Signed SEO score delta for the diff header, or null when unchanged. */
export function formatScoreDeltaFa(delta: number | null | undefined): string | null {
  if (!delta) return null;
  const magnitude = Math.abs(delta).toLocaleString("fa-IR");
  return delta > 0 ? `${magnitude}+ امتیاز سئو` : `${magnitude}- امتیاز سئو`;
}

/**
 * Author cell text. Returns null when a real user id is present, so the caller
 * resolves and renders the name itself; "سیستم" covers every write with no user
 * behind it (worker, AI, import).
 */
export function changeAuthorLabelFa(version: ContentVersionListItem): string | null {
  return version.changed_by ? null : "سیستم";
}

// ---------------------------------------------------------------------- calls

export function listVersions(params: {
  article_id: string;
  limit?: number;
  offset?: number;
}) {
  const qs = new URLSearchParams({ article_id: params.article_id });
  if (params.limit !== undefined) qs.set("limit", String(params.limit));
  if (params.offset !== undefined) qs.set("offset", String(params.offset));
  return api.get<ContentVersionListItem[]>(`/versions?${qs.toString()}`);
}

export function getVersionSummary(articleId: string) {
  const qs = new URLSearchParams({ article_id: articleId });
  return api.get<ContentVersionSummary>(`/versions/summary?${qs.toString()}`);
}

export function diffVersions(params: {
  article_id: string;
  from_version: number;
  to_version: number;
}) {
  const qs = new URLSearchParams({
    article_id: params.article_id,
    from_version: String(params.from_version),
    to_version: String(params.to_version),
  });
  return api.get<ContentVersionDiff>(`/versions/diff?${qs.toString()}`);
}

export function getVersion(versionId: string) {
  return api.get<ContentVersion>(`/versions/${versionId}`);
}

export function rollbackToVersion(
  versionId: string,
  body?: ContentVersionRollbackRequest
) {
  return api.post<ContentVersionRollbackResult>(
    `/versions/${versionId}/rollback`,
    body || {}
  );
}

File: backend/app/schemas/versions.py
"""Schemas for content versioning.

There is no create/update schema in here, and that is the point: a revision is
never written by a client. Versions are minted by `version_service` whenever the
article itself changes, so the only client-writable payload in the whole module
is the rollback note. If a client could POST a version body it could forge
history — invent an author, backdate a change, claim an edit it never made.

Two read shapes exist because a snapshot is large:

* `ContentVersionListItem` — metadata only, for the history sidebar. Dropping
  `content_markdown` / `content_html` keeps a 50-revision list from shipping a
  megabyte of duplicated article bodies over the wire.
* `ContentVersionRead` — the full snapshot, for "view this revision".

`ContentVersionDiff` carries the unified diff as a list of classified lines
rather than one blob: the UI colours each line directly, and `kind` is resolved
server-side because the raw `difflib` prefix is ambiguous — a markdown body line
that legitimately begins with `-` is indistinguishable from a deletion once the
text is concatenated.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.versions import CONTENT_CHANGE_TYPES


# ------------------------------------------------------------------ read models

class ContentVersionDiffStats(BaseModel):
    """Character and word deltas against the immediately previous version."""
    added_chars: int = 0
    removed_chars: int = 0
    added_words: int = 0
    removed_words: int = 0


class ContentVersionListItem(BaseModel):
    """One row of the history sidebar — no bodies.

    `change_summary` is included because it is short and it is the only thing
    that explains *why* a revision exists; the bodies are fetched on demand by
    `GET /versions/{version_id}`.
    """
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    website_id: UUID
    article_id: UUID
    version_number: int
    title: str
    seo_score: int
    change_type: str
    change_summary: str | None = None
    changed_by: UUID | None = None
    diff_stats: dict = Field(default_factory=dict)
    is_current: bool
    created_at: datetime


class ContentVersionRead(ContentVersionListItem):
    """A full snapshot, including both rendered forms of the body."""
    content_markdown: str
    content_html: str
    seo_metadata: dict = Field(default_factory=dict)


# ------------------------------------------------------------------------ diff

class ContentVersionDiffLine(BaseModel):
    """One line of a unified diff.

    `kind` is normalised out of the raw prefix so the frontend does not have to
    parse the first character (and does not mistake a hunk header `@@` for a
    context line, or a body line that legitimately starts with `-` for a
    deletion).
    """
    kind: str  # "context" | "added" | "removed" | "header" | "hunk"
    text: str


class ContentVersionDiff(BaseModel):
    """Line-level comparison of two revisions of the same article.

    Built with the stdlib `difflib` on `content_markdown` — the markdown is the
    authored form, so a diff over it reads like the change the person actually
    made, whereas diffing the generated HTML shows tag churn.
    """
    article_id: UUID
    from_version_number: int
    to_version_number: int
    from_created_at: datetime
    to_created_at: datetime

    # Title is compared separately: it lives outside the body but is the change
    # a reader notices first.
    title_changed: bool = False
    from_title: str
    to_title: str

    from_seo_score: int
    to_seo_score: int
    seo_score_delta: int = 0

    stats: ContentVersionDiffStats
    lines: list[ContentVersionDiffLine] = Field(default_factory=list)
    # True when the two snapshots have identical markdown — the UI shows "no
    # content change" rather than an empty diff pane that looks like a failure.
    identical: bool = False
    # Set when the diff was cut off at the line cap; the UI warns instead of
    # implying the change ended there.
    truncated: bool = False


# --------------------------------------------------------------------- summary

class ContentVersionSummary(BaseModel):
    """Header counters for the history panel."""
    article_id: UUID
    total_versions: int = 0
    current_version_number: int | None = None
    last_changed_at: datetime | None = None
    # Distinct non-null `changed_by`. Excludes worker/AI writes by construction,
    # so this is "how many people touched this", not "how many writes happened".
    contributors: int = 0


# ---------------------------------------------------------------- client writes

class ContentVersionRollbackRequest(BaseModel):
    """The only client-writable payload in this module.

    A note, nothing else. The target revision is the path parameter and the new
    version's content is copied from it server-side, so a client cannot restore
    one revision while claiming it restored another.
    """
    change_summary: str | None = Field(default=None, max_length=2000)


class ContentVersionRollbackResult(BaseModel):
    """What rollback produced: the appended version, plus where it came from."""
    restored_from_version_number: int
    new_version: ContentVersionRead


# --------------------------------------------------------------- vocabularies

class ContentVersionChangeType(BaseModel):
    """Wrapper used only to validate a change_type coming from a caller.

    Services are also driven by workers where no schema was involved, so the
    vocabulary check is repeated there; this exists for the rare router that
    accepts a change_type as a query filter.
    """
    change_type: str

    @field_validator("change_type")
    @classmethod
    def _known_change_type(cls, v: str) -> str:
        if v not in CONTENT_CHANGE_TYPES:
            raise ValueError(
                f"change_type must be one of {sorted(CONTENT_CHANGE_TYPES)}"
            )
        return v


__all__ = [
    "ContentVersionDiffStats",
    "ContentVersionListItem",
    "ContentVersionRead",
    "ContentVersionDiffLine",
    "ContentVersionDiff",
    "ContentVersionSummary",
    "ContentVersionRollbackRequest",
    "ContentVersionRollbackResult",
    "ContentVersionChangeType",
    "CONTENT_CHANGE_TYPES",
]
